pkg/datapath/linux/node.go

```diff
@@ -0,0 +1,533 @@
+	}
+
+	log.WithField("allocCIDR", oldCIDR).Debug("Deleting tunnel map entry")
+
+	if err := tunnel.TunnelMap.DeleteTunnelEndpoint(oldCIDR.IP); err != nil {
+		if !quietMode {
+			log.WithError(err).WithFields(logrus.Fields{
+				"allocCIDR": oldCIDR,
+			}).Error("Unable to delete in tunnel endpoint map")
+		}
+	}
+}
+
+func createDirectRouteSpec(CIDR *cidr.CIDR, nodeIP net.IP) (routeSpec *netlink.Route, err error) {
+	var routes []netlink.Route
+
+	routeSpec = &netlink.Route{
+		Dst: CIDR.IPNet,
+		Gw:  nodeIP,
+	}
+
+	routes, err = netlink.RouteGet(nodeIP)
+	if err != nil {
+		err = fmt.Errorf("unable to lookup route for node %s: %s", nodeIP, err)
+		return
+	}
+
+	if len(routes) == 0 {
+		err = fmt.Errorf("no route found to destination %s", nodeIP.String())
+		return
+	}
+
+	if routes[0].Gw != nil && !routes[0].Gw.IsUnspecified() {
+		err = fmt.Errorf("route to destination %s contains gateway %s, must be directly reachable",
+			nodeIP, routes[0].Gw.String())
+		return
+	}
+
+	linkIndex := routes[0].LinkIndex
+
+	// Special treatment if the route points to the loopback, lookup the
+	// local route and use that ifindex
+	if linkIndex == 1 {
+		family := netlink.FAMILY_V4
+		dst := &net.IPNet{IP: nodeIP, Mask: net.CIDRMask(32, 32)}
+		if nodeIP.To4() == nil {
+			family = netlink.FAMILY_V6
+			dst.Mask = net.CIDRMask(128, 128)
+		}
+
+		filter := &netlink.Route{
+			Table: 255, // local table
+			Dst:   dst,
+		}
+
+		routes, err = netlink.RouteListFiltered(family, filter, netlink.RT_FILTER_DST|netlink.RT_FILTER_TABLE)
+		if err != nil {
+			err = fmt.Errorf("unable to find local route for destination %s: %s", nodeIP, err)
+			return
+		}
+
+		if len(routes) == 0 {
+			err = fmt.Errorf("unable to find local route for destination %s which is routed over loopback", nodeIP)
+			return
+		}
+
+		linkIndex = routes[0].LinkIndex
+	}
+
+	routeSpec.LinkIndex = linkIndex
+
+	return
+}
+
+func installDirectRoute(CIDR *cidr.CIDR, nodeIP net.IP) (routeSpec *netlink.Route, err error) {
+	routeSpec, err = createDirectRouteSpec(CIDR, nodeIP)
+	if err != nil {
+		return
+	}
+
+	err = netlink.RouteReplace(routeSpec)
+	return
+}
+
+func (n *linuxNodeHandler) lookupDirectRoute(CIDR *cidr.CIDR, nodeIP net.IP) ([]netlink.Route, error) {
+	routeSpec, err := createDirectRouteSpec(CIDR, nodeIP)
+	if err != nil {
+		return nil, err
+	}
+
+	family := netlink.FAMILY_V4
+	if nodeIP.To4() == nil {
+		family = netlink.FAMILY_V6
+	}
+	return netlink.RouteListFiltered(family, routeSpec, netlink.RT_FILTER_DST|netlink.RT_FILTER_GW|netlink.RT_FILTER_OIF)
+}
+
+func (n *linuxNodeHandler) updateDirectRoute(oldCIDR, newCIDR *cidr.CIDR, oldIP, newIP net.IP, firstAddition, directRouteEnabled bool) error {
+	if !directRouteEnabled {
+		// When the protocol family is disabled, the initial node addition will
+		// trigger a deletion to clean up leftover entries. The deletion happens
+		// in quiet mode as we don't know whether it exists or not
+		if newCIDR != nil && firstAddition {
```

Comment:
This handles changing in configuration option. If the agent is running with direct routing and is then restarted with direct routing disabled, the first node addition of any node added will remove the old route. Blind deletions are quite costly so it's done once when the node is added for the first time.

Reply:
If I followed correctly, the callee code does a dump of routes first then attempts deletion for routes that match, rather than a blind deletion. But your comment may still apply; I don't have a sense for how expensive that is in practice.

Reply:
It is a lot more costly than what I would have guessed because it involves:
- link lookup
- route list

The route list in the kernel ist mostly sequential it seems. I measured the lookup with a couple of routes first and it was OKish but when I added 10k routes to simulate 5k nodes it became a lot more expensive, hence the optimization here.

There is obviously more we can do such as caching link lookups or listening to Netlink notifications. What is PR does is good enough for 5K nodes with some optimization potential left.